### DISCUSSION TITLE
[Idea] Cache configuration (and hooks and modules) between requests

### DISCUSSION THREAD

**Discussion by lmeyer1** [PrestaShop/PrestaShop#32827] [Ideas]
### The problem

On every request, Prestashop makes a large number of database request to load basic configuration (as configuration table, hooks, modules, shops and languages). These all are data that change only rarely, and thus are good candidates for cache.

Prestashop currently has only an array cache (store during the execution of a request), a sql cache (BO Adv. params - preformance - cache, that by the way does not really work) and a symfony filesystem cache used mainly for translations and doctrine.

We need a more flexible cache system to store datastructures loaded by Prestashop at startup time.

### The solution

There is a proof of concept https://github.com/PrestaShop/PrestaShop/pull/32824.

### Alternatives

There is no real alternative. We need to reduce the number of queries and the execution time. In the execution time, initialization (mainly in `config.inc.php`) is particularly important, because it is executed at every request.

### Additional context

Using Prestashop Profiler with fix https://github.com/PrestaShop/PrestaShop/pull/32754, I benchmarked the home page of our shop with and without https://github.com/PrestaShop/PrestaShop/pull/32824.

![2023-06-05_113816](https://github.com/PrestaShop/PrestaShop/assets/7520695/ad3419b0-9635-4ded-930d-f4e6c282a9e0)

Here are the results: 

Without optimization
------
|  | config time [ms] | mem [MB] | peak mem [MB] | queries | load time [ms]
| - | ----- | --- | --- | -- | ------
| mean (5 measurements) |  74.05  |  3.90  |  5.00  |  92  |  176.66 
| std dev |  9.33  |  -    |  -    |  -    |  15.51 

With commit https://github.com/PrestaShop/PrestaShop/pull/32824/commits/e0e9d4c860bfb91f21605f0c5678ffa760f1099c
-----
|  | config time [ms] | mem [MB] | peak mem [MB] | queries | load time [ms]
| - | ----- | --- | --- | -- | ------
| mean (5 measurements) |  71.49  |  3.48  |  4.70  |  92  |  175.45 
| std dev |  5.17  |  -  |  -    |  -    |  11.85 
| improvement | -2.56  | -0.42  | -0.30  |  -    | -1.20 
| in %  | -3% | -11% | -6% | 0% | -1%
| significative improvement | no | yes | yes |  | no

Removing the deprecated property `Configuration::$_cache` is reducing memory footprint of a significative amount, but does not improve speed.

Up to commit https://github.com/PrestaShop/PrestaShop/pull/32824/commits/5c708f6f73676d4053cf29f4d6d47a6e6602b6ab
----
|  | config time [ms] | mem [MB] | peak mem [MB] | queries | load time [ms]
| - | ----- | --- | --- | -- | ------
| mean |  32.18  |  3.36  |  3.70  |  52  |  123.53 
| std dev |  2.61  |  -  |  -    |  -    |  5.52 
| improvement | -41.88  | -0.54  | -1.30  | -40  | -53.13 
| en %  | -57% | -14% | -26% | -43% | -30%
| significative improvement | yes | yes | yes | yes | yes

This PR gives substantial improvement on load time. The `config time` is reduced by more than half, and we have 40 queries less, that represents 40% on the page used for the benchmark.

On our production site, the average number of queries / request is 175. 40 queries less would represent a overall reduction of 23%. For load time, we have an average of 414ms. An improvement of 41ms represents 10%, which is not neglectable.

### Do you plan to work on this subject?

- [ ] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [X] I'm willing to submit a Pull Request that implements this feature.
- [X] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

**Comment by matks** [comment]
You're 100% right

> (as configuration table, hooks, modules, shops and languages)

However these are critical data, so we need to be very careful about how the data is cached and **refreshed**

**Comment by lmeyer1** [comment]
@matks 
I fear, that we can never be sure to catch all events that need a cache refresh.
Just thinking about hook aliases that are added when upgrading PS, all the options in the BO to configure module hooks, etc.
While it is possible, yet difficult to catch all interactions in the BO, it is not possible to have the cache react to data that is changed through SQL.

**Comment by cruftex** [comment]
@matks, @lmeyer1 

> However these are critical data, so we need to be very careful about how the data is cached and refreshed
>> I fear, that we can never be sure to catch all events that need a cache refresh.

The basic problem is, as soon as you add caching you are getting concurrency issues that you usually don't have in state less PHP applications. I also commented here https://github.com/PrestaShop/PrestaShop/pull/32824#issuecomment-1596883840, and duplicating this to have the essentials here:

I can see this execution sequence:

- update database
- invalidate cache

This is not working correctly in general and contains a race. Consider a second process doing this:

- check for cached item is available
- generate the cache item, if needed
- write cached item to cache

So there is this potential serialised execution sequence:

- P2: check for cached item is available
- P2: generate the cache item, if needed
- P1: update database
- P1: invalidate cache
- P2: write cached item to cache

To mitigate this, you need CAS or locking support in the cache. I am not a PHP guy, but it seems the Symphony cache interface is not providing this. That means you cannot use Symphony cache for data that needs to be kept consistent :(

**Comment by lmeyer1** [comment]
> The basic problem is, as soon as you add caching you are getting concurrency issues that you usually don't have in state less PHP applications.

These data (except configuration) changes extremely rarely, and the cache exclusively invalidates by expiration. For all but configuration, concurrency is impossible.  For the configuration cache item, while a theoretical possibility, this seems to me impossible in real life. Impossible here means extremely low probability.
See https://github.com/PrestaShop/PrestaShop/pull/32824#issuecomment-1598266203 for more details.

**Comment by cruftex** [comment]
I commented also in the PR.

Unfortunately, on a high load server with thousand requests per second, there is a high chance that an overlap happens. Symfony is already doing something to avoid the cache stampede and using a callback, so maybe it is less likely depending on the actual cache implementation used.
 
I am not sure what you mean with "cache exclusively invalidates by expiration", since your solution invalidates explicitly after a data change.

Anyhow, I think it would be better to discuss how to make the cache consistent, and not whether having consistent configuration data is a good thing to have :)

One approach could be this:

- On a data write, instead of invalidating the cache, you could write the actual value to the cache.
- On a data read, if you want to initially populate the cache, you must do the cache update only if the item is still not present. With Redis that would be the operation SETNX / SET if Not eXists. With ACPU it is apcu_add.

That fixes the race of concurrent write and reads. On concurrent writes the cache can become still inconsistent. That would be solvable as well. Anyhow, all sorts of bad things can happen when several admins change something at the same time.

It is important, that the cache can not be consistent across multiple keys. So e.g. the hooks always need to be cached as one entry.
